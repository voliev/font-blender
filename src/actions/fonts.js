import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
