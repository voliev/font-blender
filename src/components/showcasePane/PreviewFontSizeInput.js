import React, { Fragment } from "react";
import debounce from "lodash.debounce";
import PropTypes from "prop-types";
import "./styles/PreviewFontSizeInput.css";

const PreviewFontSizeInput = ({ fontSizeChangeHandler, value }) => (
  <Fragment>
    <label className="PreviewFontSizeInput label" htmlFor="preview-font-size">
      Font size (px)
    </label>
    <input
      className="PreviewFontSizeInput input"
      id="preview-font-size"
      type="number"
      name="font-size"
      min="1"
      max="100"
      value={value}
      onChange={debounce(fontSizeChangeHandler, 100, {
        leading: true
      })}
    />
  </Fragment>
);

PreviewFontSizeInput.propTypes = {
  fontSizeChangeHandler: PropTypes.func.isRequired,
  value: PropTypes.number.isRequired
};

export default PreviewFontSizeInput;
