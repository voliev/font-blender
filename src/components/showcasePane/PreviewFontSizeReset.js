import React from "react";
import PropTypes from "prop-types";
import "./styles/PreviewFontSizeReset.css";

const PreviewFontSizeReset = ({ resetFontSizeHandler }) => (
  <button
    className="PreviewFontSizeReset btn-blue-rect"
    onClick={resetFontSizeHandler}
    value="reset"
    title="Reset font size"
  >
    Reset
  </button>
);

PreviewFontSizeReset.propTypes = {
  resetFontSizeHandler: PropTypes.func.isRequired
};

export default PreviewFontSizeReset;
