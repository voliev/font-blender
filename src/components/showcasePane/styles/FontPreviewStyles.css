.FontPreviewStyles .heading {
  border-bottom: 1px solid var(--light-grey);
  padding-bottom: var(--space-md);
}

.FontPreviewStyles .font-size-controls {
  align-items: center;
  display: flex;
}
