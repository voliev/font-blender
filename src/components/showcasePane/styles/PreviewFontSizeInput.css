.PreviewFontSizeInput {
  align-items: center;
  display: flex;
}

.PreviewFontSizeInput.label,
.PreviewFontSizeInput.input {
  margin-right: var(--space-md);
}

.PreviewFontSizeInput.label {
  flex: 0 0 7rem;
}

.PreviewFontSizeInput.input {
  flex: auto;
}
