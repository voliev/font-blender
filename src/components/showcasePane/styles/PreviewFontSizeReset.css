.PreviewFontSizeReset {
  height: 2rem;
  padding: var(--space-xs) var(--space-md);
}
