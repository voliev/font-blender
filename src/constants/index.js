// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
