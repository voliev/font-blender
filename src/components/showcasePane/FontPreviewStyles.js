import React, { Component } from "react";
import PropTypes from "prop-types";
import { DEFAULT_FONT_SIZE } from "../../constants";
import PreviewFontSizeInput from "./PreviewFontSizeInput";
import PreviewFontSizeReset from "./PreviewFontSizeReset";
import FontPreviewVariants from "./FontPreviewVariants";
import { numberInputChangeHandler } from "../../utils";
import "./styles/FontPreviewStyles.css";

class FontPreviewStyles extends Component {
  constructor() {
    super();

    this.state = {
      fontSize: DEFAULT_FONT_SIZE
    };

    this.handleFontSizeChange = this.handleFontSizeChange.bind(this);
    this.handleResetFontSize = this.handleResetFontSize.bind(this);
  }

  handleFontSizeChange(e) {
    numberInputChangeHandler(e.target.value, 16, 1, 100, fontSize => {
      this.setState({ fontSize });
    });
  }

  handleResetFontSize() {
    this.resetFontSize();
  }

  resetFontSize() {
    this.setState({
      fontSize: DEFAULT_FONT_SIZE
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.family !== this.props.family) {
      this.resetFontSize();
    }
  }

  render() {
    const { family, category, variants } = this.props;

    const { fontSize } = this.state;

    return (
      <div className="FontPreviewStyles">
        <h3 className="heading">Styles</h3>
        <div className="font-size-controls">
          <PreviewFontSizeInput
            fontSizeChangeHandler={this.handleFontSizeChange}
            value={fontSize}
          />
          <PreviewFontSizeReset
            resetFontSizeHandler={this.handleResetFontSize}
          />
        </div>
        <FontPreviewVariants
          category={category}
          family={family}
          fontSize={fontSize}
          variants={variants}
        />
      </div>
    );
  }
}

FontPreviewStyles.propTypes = {
  category: PropTypes.string.isRequired,
  family: PropTypes.string.isRequired,
  variants: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired
};

export default FontPreviewStyles;
