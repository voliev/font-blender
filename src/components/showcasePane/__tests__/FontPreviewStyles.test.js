import React from "react";
import { mount, shallow } from "enzyme";
import FontPreviewStyles from "../FontPreviewStyles";
import PreviewFontSizeInput from "../PreviewFontSizeInput";
import PreviewFontSizeReset from "../PreviewFontSizeReset";
import { DEFAULT_FONT_SIZE } from "../../../constants";
import fonts from "../../../fixtures/fonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const font = fonts["Roboto"];
  const props = {
    family: font.family,
    category: font.category,
    variants: font.variants,
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<FontPreviewStyles {...props} />)
  };
};

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should set default font size on initial render", () => {
  const { wrapper } = setup();

  expect(wrapper.state("fontSize")).toBe(DEFAULT_FONT_SIZE);
});

it("should receive corresponding props", () => {
  const { wrapper, category, family, variants } = setup(mount);

  expect(typeof wrapper.prop("category")).toBe("string");
  expect(wrapper.prop("category")).toBe(category);

  expect(typeof wrapper.prop("family")).toBe("string");
  expect(wrapper.prop("family")).toBe(family);

  expect(typeof wrapper.prop("variants")).toBe("object");
  expect(wrapper.prop("variants").hasOwnProperty("length")).toBeTruthy();
  expect(wrapper.prop("variants")).toBe(variants);
});

it("should handle font size change", () => {
  const { wrapper } = setup();
  const event = {
    target: {
      value: "18"
    }
  };

  wrapper.find(PreviewFontSizeInput).prop("fontSizeChangeHandler")(event);

  expect(wrapper.state("fontSize")).toBe(18);
});

it("should handle font size reset", () => {
  const { wrapper } = setup();
  const event = {
    target: {
      value: "18"
    }
  };

  wrapper.find(PreviewFontSizeInput).prop("fontSizeChangeHandler")(event);

  expect(wrapper.state("fontSize")).toBe(18);

  wrapper.find(PreviewFontSizeReset).prop("resetFontSizeHandler")();

  expect(wrapper.state("fontSize")).toBe(DEFAULT_FONT_SIZE);
});

it("should reset font size on font family change", () => {
  const { wrapper } = setup();
  const event = {
    target: {
      value: "18"
    }
  };

  wrapper.find(PreviewFontSizeInput).prop("fontSizeChangeHandler")(event);

  expect(wrapper.state("fontSize")).toBe(18);

  const newFont = fonts["Open Sans"];

  wrapper.setProps({
    category: newFont.category,
    family: newFont.family,
    variants: newFont.variants
  });

  expect(wrapper.state("fontSize")).toBe(DEFAULT_FONT_SIZE);
});
