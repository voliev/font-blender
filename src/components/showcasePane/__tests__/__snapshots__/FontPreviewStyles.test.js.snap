// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="FontPreviewStyles"
>
  <h3
    className="heading"
  >
    Styles
  </h3>
  <div
    className="font-size-controls"
  >
    <PreviewFontSizeInput
      fontSizeChangeHandler={[Function]}
      value={16}
    />
    <PreviewFontSizeReset
      resetFontSizeHandler={[Function]}
    />
  </div>
  <FontPreviewVariants
    category="sans-serif"
    family="Roboto"
    fontSize={16}
    variants={
      Array [
        "100",
        "100italic",
        "300",
        "300italic",
        "regular",
        "italic",
        "500",
        "500italic",
        "700",
        "700italic",
        "900",
        "900italic",
      ]
    }
  />
</div>
`;
